realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -504,4 +514,14 @@ private int getTotalGlobalRefCount() {
 
         return totalRefCount;
     }
+
+    /**
+     * If a Realm instance is GCed but `Realm.close()` is not called before, we still want to track the cache for
+     * debugging. Adding them to the list to keep the strong ref of the cache to prevent the cache gets GCed.
+     */
+    void leak() {
+        if (!isLeaked.getAndSet(true)) {
+            leakedCaches.add(this);
```

Comment:
hmmm, this case is interesting, actually the behaviour is the same even before this change:
We are holding `ThreadLocal` to the Realm instance, the `ThreadLocal` is holding a weak ref to it underneath. So when above happens, the 3rd line (`Realm r = Realm.getDefaultInstance();`) will just return a `null` instance since the weak ref to the Realm instance has been just released. 

So I think it is better we just throw exception when we detect this leak to prevent user from getting a `null` instance?